Compiler team 2022 ambitions (#935)
rust-lang/blog.rust-lang.org: posts/inside-rust/2022-02-22-compiler-team-ambitions-2022.md

[Debugging concrete]: #Debugging-Initiatives
[Debugging aspirational]: #Debugging-Aspirations
[Debugging]: #Debugging-Object-Code
[Faster concrete]: #Compilation-Time-Initiatives
[Faster aspirational]: #Faster-Compilation-Aspirations
[Expressiveness concrete]: #Language-Expressiveness-Initiatives
[Expressiveness aspirational]: #Expressiveness-Aspirations

### Contributor Workflow (🛠️)

Finally, improving the Compiler Contributor Workflow means technology enhancements that benefit people maintaining and extending the Rust compiler itself.

(We also make non-technical enhancements, such as changes to our social processes, but this document focuses on technology.)

[Librarification Initiatives]: #Librarification-Initiatives
[Librarification Aspirations]: #Librarification-Aspirations

[Compiler Backend]: #Compiler-Backend
[Review Queue Crew]: #Review-Queue-Crew

Category  | [Concrete Initiatives] |  [Aspirations]
----------|---------------------|-----------
I-unsound (🦀) | [Initiatives][I-unsound Issues]      |
Async Rust (🦀, 👩‍💻)| [Initiatives][Async Initiatives]     |
Debugging (🦀, 👩‍💻)| [Initiatives][Debugging Initiatives] | [Aspirations][Debugging Aspirations]

> Faster Builds (👩‍💻, 🛠️) | [Initiatives][Faster Builds Initiatives] | [Aspirations][Faster Builds Aspirations]

Author: I'm working on fixing the broken links in this table now.

In addition to the LLVM and Cranelift backends, there is also a new backend under development that uses `libgccjit` from GCC (which, as many have clarified, is usable for ahead-of-time as well as just-in-time compilation). This backend enables Rust to target more platforms that are not supported by LLVM.

If you are interested in helping out with this project, reach out to [antoyo] (supported via [sponsorship](https://github.com/sponsors/antoyo)) and [bjorn3].

### Diagnostics Aspirations (👩‍💻)

The Rust compiler has pretty good diagnotics. But the good news is, there's a [full employment theorem](https://en.wikipedia.org/wiki/Full_employment_theorem) for diagnostics engineers which is supported by the 1,500+ [open diagnostics issues](https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3AA-diagnostics) we have.

Diagnostics improvements are an *excellent* first step for learning about how to contribute to the Rust compiler. If you're interested in helping out but don't have any idea where to start, fixing diagnostic bugs is a great jumping off point, and you can reach out to [estebank], at Amazon Web Services, to find out more about how to help.

> ## Conclusion

Reviewer: I'm planning to rewrite this tomorrow.

<!-- Aaron1011: no affiliation -->
[bjorn3]: https://github.com/bjorn3
<!-- [bjorn3 zulip PM]: https://rust-lang.zulipchat.com/#narrow/pm-with/133247-user133247 --> <!-- @**bjorn3**  -->
<!-- bjorn3 donation page: https://liberapay.com/bjorn3 -->
[cjgillot]: https://github.com/cjgillot
<!-- [cjgillot zulip PM]: https://rust-lang.zulipchat.com/#narrow/pm-with/248906-user248906 --> <!-- @**cjgillot**  -->
<!-- no response from cjgillot re affiliation yet -->
[davidtwco]: https://github.com/davidtwco
<!-- [davidtwco zulip PM]: https://rust-lang.zulipchat.com/#narrow/pm-with/248906-user248906 --> <!-- @**davidtwco**  -->
<!-- davidtwco affiliation: "Huawei R&D UK"-->
[estebank]: https://github.com/estebank
<!-- [estebank zulip PM]: https://rust-lang.zulipchat.com/#narrow/pm-with/119031-user119031 --> <!-- @**Esteban Küber** -->
<!-- estebank affiliation: AWS -->
[lcnr]: https://github.com/lcnr
<!-- [lcnr zulip PM]: https://rust-lang.zulipchat.com/#narrow/pm-with/216206-user216206 --> <!-- @**lcnr** -->
<!-- lcnr sponsorship: https://lcnr.de/funding/ -->
[michaelwoerister]: https://github.com/michaelwoerister

> <!-- [mw zulip PM]: https://rust-lang.zulipchat.com/#narrow/pm-with/124287-user124287 --> <!-- @**michaelwoerister** -->

Author: ```suggestion
<!-- [michaelwoerister zulip PM]: https://rust-lang.zulipchat.com/#narrow/pm-with/124287-user124287 --> <!-- @**mw** -->
```

(The idea here was to put the zulip user names in the comment, for ease of lookup/verification when adding new references to the person elsewhere.)

> <!-- mw affiliation: MS -->

Author: ```suggestion
<!-- michaelwoerister affiliation: MS -->
```

#### async traits

Rust today does not allow `async fn` in a trait, so Async Rust code usually ends up with components that are too tightly coupled; one cannot write reusable, general-purpose libraries without using workarounds like `#[async_trait]` that impose hidden costs. [nikomatsakis], from Amazon Web Services, and [tmandry], from Google, are driving the [async fn in traits initiative](https://github.com/rust-lang/async-fundamentals-initiative/issues/5), which will unlock the ability to write `async` methods in traits, natively. They are `@**nikomatsakis**` and `@**tmandry**` [on zulip].

#### async crashdump dissection

[michaelwoerister], from Microsoft, is driving the [async crashdump initiative](https://rust-lang.github.io/async-crashdump-debugging-initiative/), which will enable developers to understand the control-flow stacks encoded in crashdumps for their async Rust programs. He is `@**mw**` [on zulip].

There is a ton of other work being done in the Async Rust space. Check out the [Async Vision web site](https://rust-lang.github.io/wg-async/welcome.html) for more information.

### Debugging Initiatives (🦀)

> [wesleywiser], from Microsoft, and [pnkfelix] are spinning up a wg-debugging working group. It will cover at least the following sub-items: improving Rust's debuginfo quality ([mw], [wesleywiser]), supporting split debuginfo ([davidtwco], from Huawei R&D UK), and better integration with trace-based debuggers like `rr` ([pnkfelix]). They are `@**Wesley Wiser**`, `@**pnkfelix**`, `@**mw**` and `@**davidtwco**` [on zulip].

Author: ```suggestion
[wesleywiser], from Microsoft, and [pnkfelix] are spinning up a wg-debugging working group. It will cover at least the following sub-items: improving Rust's debuginfo quality ([michaelwoerister], [wesleywiser]), supporting split debuginfo ([davidtwco], from Huawei R&D UK), and better integration with trace-based debuggers like `rr` ([pnkfelix]). They are `@**Wesley Wiser**`, `@**pnkfelix**`, `@**mw**` and `@**davidtwco**` [on zulip].
```

(just adding one rewrite you missed.)

### Diagnostics Aspirations (👩‍💻)

The Rust compiler has pretty good diagnotics. But the good news is, there's a [full employment theorem](https://en.wikipedia.org/wiki/Full_employment_theorem) for diagnostics engineers which is supported by the 1,500+ [open diagnostics issues](https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3AA-diagnostics) we have.

Diagnostics improvements are an *excellent* first step for learning about how to contribute to the Rust compiler. If you're interested in helping out but don't have any idea where to start, fixing diagnostic bugs is a great jumping off point, and you can reach out to [estebank], at Amazon Web Services, to find out more about how to help. They are `@**Esteban Küber**` [on zulip].

## Conclusion

Reading over this list, the number of items on it seems quite daunting! We believe these initiatives will provide the highest impact to the Rust community by helping to fulfill Rust's promise, delighting Rust developers and improving our contributor workflows and aligns well with the results of the [2021 Rust Survey](https://blog.rust-lang.org/2022/02/15/Rust-Survey-2021.html).

> While we think we will be able to make signficant progress on these initiatives this year, project estimation is a difficult and inexact science, especially for open source projects. What we will achieve is ultimately a result of who decides to contribute. Our aspirational goals are currently just that: aspriations.

Author: ```suggestion
While we think we will be able to make signficant progress on these initiatives this year, project estimation is a difficult and inexact science, especially for open source projects. What we will achieve is ultimately a result of who decides to contribute. Our aspirational goals are currently just that: aspirations.
```

i can't believe how many times I made this same typo when I was writing

 
The Rust compiler's end-to-end latency is known to be a problem.
 
[lqd], sponsored by the Internet Security Research Group, is dedicating the majority of 2022 to working on this, partnering with Rust's compiler-performance working group as well as performance experts like [nnethercote] (from Futurewei Technologies). [lqd] has their own [living document](https://hackmd.io/3Dp68rTDSpWvRDfWF6lbMw?view) that lists areas under investigation, and [nnethercote] has a [roadmap under development](https://hackmd.io/YJQSj_nLSZWl2sbI84R1qA). They are `@**lqd**` and `@**nnethercote**` [on zulip].

[ISRG]: https://www.abetterinternet.org/

### Expressiveness Initiatives (👩‍💻, 🦀)

A common refrain we hear is: "I need feature X, but it's not implemented in rustc or stable." 
In Rust, we use an open Request-for-Comment (RFC) process for designing new features. Currently, we have [this set of RFCs approved][RFC tracking issue list]; here are some imporant features with dedicated owners that we expect forward movement on.

[RFC tracking issue list]: https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3AC-tracking-issue++label%3AB-RFC-approved+

> Generic Associated Types, or [GATs](https://github.com/rust-lang/generic-associated-types-initiative/issues/4), are an ongoing effort owned by [jackh726]. GATs have many applications, such as traits whose associated types have lifetimes tied to the local borrowing of the receiver type ([e.g. `StreamingIterator`][GAT-motivation]).

Author: ```suggestion
Generic Associated Types, or [GATs](https://github.com/rust-lang/generic-associated-types-initiative/issues/4), are an ongoing effort owned by [jackh726]. GATs have many applications, such as traits whose associated types have lifetimes tied to the local borrowing of the receiver type ([e.g. `LendingIterator`][GAT-motivation]).
```

If you want to help out here, please reach out to [pnkfelix] and [wesleywiser].

### Faster Builds Aspirations (👩‍💻, 🛠️)

#### Parallel Compilation

Parallel Compilation is one avenue for improving compiler performance. It is also a very complex area, especially when it comes to the tradeoff of how much of a hit one is willing to take on single core builds in order to enable more parallel computation. We already parallelize our LLVM invocations, but the parallelization of the rest of the compiler remains in an experimental state. This is an area we think needs long-term collaborative effort with the compiler team. We do not expect to deliver a solution here this year.

If you want to discuss more with us about past attempts and ideas for the future, please reach out to [pnkfelix] and [wesleywiser].

#### Incremental Compilation Aspirations

Incremental compilation performance and stability are both ongoing concerns to the team. We *know* there is significant room to improve the effectiveness of incremental compilation, in terms of reducing the amount of redundant work done by successive `rustc` invocations.

> In addition, there is a significant amount of work that could be done to improve our testing infrastructure for incremental compilation which does not require deep knowledge of the compiler.

Author: ```suggestion
In addition, there is a significant amount of work that could be done to improve our testing infrastructure for incremental compilation which does not require deep knowledge of the compiler. We have had to disable and subsequently reenable incremental compilation on the stable release; we want to expand our validation strategies so that we get alerted to problems in incremental compilation well before they come close to the stable channel.
```

> In addition, there is a significant amount of work that could be done to improve our testing infrastructure for incremental compiliation which does not require deep knowledge of the compiler.

Reviewer 2: hi it's me again with them typos 🤡 

```suggestion
In addition, there is a significant amount of work that could be done to improve our testing infrastructure for incremental compilation which does not require deep knowledge of the compiler.
```

If you want to learn more, reach out to [cjgillot] and [Aaron Hill].

#### Inter-crate Sharing Aspirations

nnethercote has noted that there may be opportunities
to improve end-to-end compilation time for multi-crate builds by identifying redundant activity that can be shared between builds of distinct crates. (For example, the metadata from libstd is read and decoded on every single crate compile.)

If you are interested in exploring this idea further, reach out to [nnethercote] and [lqd].

### Expressiveness Aspirations (🦀, 👩‍💻)

const generics and const eval are making steady progress. There are a *lot* of feature flags, which implies there's a lot of knobs that could be turned on and off.

What we can probably use the most help with is in identifying what subset of the features we should be striving to stabilize in order to unlock specific use cases for Rust developers.

> So, if you or your team is enthuastically awaiting const generics or const eval, reach out to [lcnr] and [oli-obk].

Reviewer 2: tiny typo 

```suggestion
So, if you or your team is enthusiastically awaiting const generics or const eval, reach out to [lcnr] and [oli-obk].
```

[Rustc Dev Guide]: https://rustc-dev-guide.rust-lang.org/
[Zulip]: https://rust-lang.zulipchat.com/
[on zulip]: https://rust-lang.zulipchat.com/#narrow/stream/122652-new-members/topic/compiler.202022

#### What do I do if I'm interested in compiler development but have no experience in compilers?

This is not a problem! Many members of our community learned about compilers by working on rustc, and we encourage others to do so as well. You can start by reading the [Rustc Dev Guide] and by joining us on [Zulip]. You may also benefit from watching the RustConf 2021 presentation on [Contributing to the Compiler] by [estebank].

[Contributing to the Compiler]: https://www.youtube.com/watch?v=vCODCbUSA_w

In addition, there are areas in this project where people without compiler expertise can have impact. For example, as mentioned in the [Performance Dashboard](#Performance-Dashboard) section, some of our internal tools could use some web front-end work.

> ### How can I contact an item's owners or sponsor their work on Rust?

Author: ```suggestion
#### How can I contact an item's owners or sponsor their work on Rust?
```

make this question's format match the other FAQ entries.

This table lists the item owners mentioned above, their [Zulip] username and if they are accepting sponsorships to help them work on Rust:

Owner | Zulip Username | Accepting sponsorships?
-|-|-
[Aaron Hill] | `@Aaron Hill` | No
[antoyo] | `@antoyo` | Yes: [GitHub Sponsors](https://github.com/sponsors/antoyo)
[apiraino] | `@apiraino` | No

> [bjorn3] | `@bjorn3` | Yes: [Librapay](https://liberapay.com/bjorn3)

Reviewer 2: Tiny typo

```suggestion
[bjorn3] | `@bjorn3` | Yes: [Liberapay](https://liberapay.com/bjorn3)
```

[cjgillot] | `@cjgillot` | No
[davidtwco] | `@davidtwco` | No: works on Rust at Huawei R&D UK
[estebank] | `@Esteban Küber` | No: works on Rust at Amazon Web Services
[jackh726] | `@Jack Huey` | No
[jswrenn] | `@Jack Wrenn` | No: works on Rust at Amazon Web Services

> [lcnr] | `@lcnr` | Yes: https://lcnr.de/funding/

Author: ```suggestion
[lcnr] | `@lcnr` | Yes: [https://lcnr.de/funding/](https://lcnr.de/funding/)
```

(without this, the url doen't render as a hyper link, at least in the blog post rendered preview )

#### I did not see any mention of monadic burritos (or other non-Rust language feature); why is that not part of your plan?

> The scope of this doc is largely restricted to Compiler Team issues. Language design work is done by the Language Team. You can reach out to them about their initiatives for this year and beyond.

Author: technically T-lang does refer to itself as "the language design team", since one can argue that "language" encompasses design *and implementation* (i.e. T-compiler is the latter).

But I don't actually care, and I think the text is indeed improved with your change (it sounds less snarky with your wording, even if it is less quote-unquote precise).

Reviewer: Ah, ok good to know, thanks! The repeated use of "language design" in such close proximity was bothering me and rust-lang.org/governance just refers to it as the "language team" so that's what I went with.

Reviewer: Is the lang team planning to write some kind of blog post like this? If so, perhaps we can say something like
 
```suggestion
The scope of this doc is largely restricted to Compiler Team issues. The Language Team is planning to write more about their initiatives for this year and beyond in another post. Stay tuned for that!
```
